java/vector/src/main/java/org/apache/arrow/vector/VectorSchemaRoot.java

```diff
@@ -312,5 +315,63 @@ public VectorSchemaRoot slice(int index, int length) {
     return new VectorSchemaRoot(sliceVectors);
   }
 
+  /**
+   * Determine if two record batches are exactly equal.
+   */
+  public boolean equals(VectorSchemaRoot other) {
+    if (other == null) {
+      return false;
+    }
+
+    if (!this.schema.equals(other.schema)) {
+      return false;
+    }
+
+    if (this.rowCount != other.rowCount) {
+      return false;
+    }
+
+    for (int i = 0; i < fieldVectors.size(); i++) {
+      FieldVector vector = fieldVectors.get(i);
+      FieldVector otherVector = other.fieldVectors.get(i);
+      if (!VectorEqualsVisitor.vectorEquals(vector, otherVector)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Determine if two record batches are approximately equal.
+   */
+  public boolean approxEquals(VectorSchemaRoot other) {
+    if (other == null) {
+      return false;
+    }
+
+    if (!this.schema.equals(other.schema)) {
+      return false;
+    }
+
+    if (this.rowCount != other.rowCount) {
+      return false;
+    }
+
+    for (int i = 0; i < fieldVectors.size(); i++) {
+      FieldVector vector = fieldVectors.get(i);
+      FieldVector otherVector = other.fieldVectors.get(i);
+      if (vector.getValueCount() != otherVector.getValueCount()) {
+        return false;
+      }
+      ApproxEqualsVisitor visitor = new ApproxEqualsVisitor(vector, otherVector);
+      Range range = new Range(0, 0, vector.getValueCount());
+      if (!visitor.rangeEquals(range)) {
```

Comment:
Is it possible to reuse the visitor and range objects?

Reply:
Thanks, visitors are not reusable. I updated this PR and make Range reusable.